File: code/scenes/src/main/java/org/fbme/scenes/controllers/scene/FullEditorBackgroundDragFacility.java
package org.fbme.scenes.controllers.scene;

import jetbrains.mps.nodeEditor.EditorComponent;
import jetbrains.mps.openapi.editor.cells.EditorCell;
import jetbrains.mps.openapi.editor.style.Style;
import org.fbme.scenes.controllers.DragEventHandler;
import org.fbme.scenes.controllers.LayoutUtil;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;

public class FullEditorBackgroundDragFacility {

    @NotNull
    private final SceneEditor myScene;

    @NotNull
    private final EditorCell myCell;

    @NotNull
    private final JViewport myViewport;

    private final Layer myLayer;

    public FullEditorBackgroundDragFacility(@NotNull SceneEditor scene, @NotNull EditorCell cell, @NotNull JViewport viewport, Layer backgroundLayer) {
        myScene = scene;
        myViewport = viewport;
        myCell = cell;
        myLayer = backgroundLayer;
        myScene.addInitializer(new MyInitializer());
    }

    public class MyInitializer implements SceneInitializer {
        private final MouseWheelListener myMouseWheelListener = new MouseWheelListener() {
            @Override
            public void mouseWheelMoved(MouseWheelEvent event) {
                if (event.isMetaDown()) {
                    int unitsToScroll = event.getUnitsToScroll();
                    Style style = myCell.getStyle();
                    int currentFontSize = LayoutUtil.getFontSize(style);
                    int newLineSize = currentFontSize + unitsToScroll;
                    int cellEventX = event.getX();
                    int cellEventY = event.getY();
                    Point editorShift = myViewport.getViewPosition();
                    int modelEventX = (int) ((cellEventX - editorShift.x) / scale(currentFontSize));
                    int modelEventY = (int) ((cellEventY - editorShift.y) / scale(currentFontSize));
                    int newCellEventX = (int) (scale(newLineSize) * modelEventX) + editorShift.x;
                    int newCellEventY = (int) (scale(newLineSize) * modelEventY) + editorShift.y;
                    LayoutUtil.setFontSize(style, newLineSize);

                    Rectangle viewRect = myViewport.getViewRect();
                    viewRect.x += newCellEventX - cellEventX;
                    viewRect.y += newCellEventY - cellEventY;
                    ((JComponent) myViewport.getView()).scrollRectToVisible(viewRect);

                    EditorComponent component = (EditorComponent) myScene.getContext().getEditorComponent();
                    component.relayout();
                    component.revalidate();
                }
            }

            private float scale(int lineSize) {
                return ((float) lineSize) / 100;
            }
        };
        private final DragEventListener myDragListener = event -> {
            if (event.awt.isMetaDown()) {
                event.consume(new MyViewpointModifier(event.awt.getX(), event.awt.getY()));
            }
        };


        @Override
        public void onAdd() {
            myScene.addMouseWheelListener(myLayer, myMouseWheelListener);
            myScene.addDragListener(myLayer, myDragListener);
        }

        @Override
        public void onRemove() {
            myScene.removeMouseWheelListener(myMouseWheelListener);
            myScene.removeDragListener(myDragListener);

        }
    }

    public class MyViewpointModifier implements DragEventHandler {

        private final int originX;
        private final int originY;

        public MyViewpointModifier(int originX, int originY) {
            this.originX = originX;
            this.originY = originY;
        }

        @Override
        public void dragTo(int x, int y) {
            Rectangle viewRect = myViewport.getViewRect();
            viewRect.x += originX - x;
            viewRect.y += originY - y;
            Object o = myViewport.getView();
            ((JComponent) o).scrollRectToVisible(viewRect);
        }
    }

}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/scene/FullEditorBackgroundDragFacility.kt
package org.fbme.scenes.controllers.scene

import jetbrains.mps.nodeEditor.EditorComponent
import jetbrains.mps.openapi.editor.cells.EditorCell
import org.fbme.scenes.controllers.DragEventHandler
import org.fbme.scenes.controllers.LayoutUtil
import java.awt.event.MouseWheelEvent
import java.awt.event.MouseWheelListener
import javax.swing.JComponent
import javax.swing.JViewport

class FullEditorBackgroundDragFacility(
    private val scene: SceneEditor,
    private val cell: EditorCell,
    private val viewport: JViewport,
    private val layer: Layer
) {
    inner class MyInitializer : SceneInitializer {
        private val mouseWheelListener: MouseWheelListener = object : MouseWheelListener {
            override fun mouseWheelMoved(event: MouseWheelEvent) {
                if (MouseEvents.isMetaDown(event)) {
                    val unitsToScroll = event.unitsToScroll
                    val style = cell.style
                    val currentFontSize = LayoutUtil.getFontSize(style)
                    val newLineSize = currentFontSize + unitsToScroll
                    val cellEventX = event.x
                    val cellEventY = event.y
                    val editorShift = viewport.viewPosition
                    val modelEventX = ((cellEventX - editorShift.x) / scale(currentFontSize)).toInt()
                    val modelEventY = ((cellEventY - editorShift.y) / scale(currentFontSize)).toInt()
                    val newCellEventX = (scale(newLineSize) * modelEventX).toInt() + editorShift.x
                    val newCellEventY = (scale(newLineSize) * modelEventY).toInt() + editorShift.y
                    LayoutUtil.setFontSize(style, newLineSize)
                    val viewRect = viewport.viewRect
                    viewRect.x += newCellEventX - cellEventX
                    viewRect.y += newCellEventY - cellEventY
                    (viewport.view as JComponent).scrollRectToVisible(viewRect)
                    val component = scene.editorContext.editorComponent as EditorComponent
                    component.relayout()
                    component.revalidate()
                }
            }

            private fun scale(lineSize: Int): Float {
                return lineSize.toFloat() / 100
            }
        }
        private val dragListener = object : DragEventListener {
            override fun onDragStarted(event: DragEvent) {
                if (MouseEvents.isMetaDown(event.awt)) {
                    event.consume(MyViewpointModifier(event.awt.x, event.awt.y))
                }
            }
        }

        override fun onAdd() {
            scene.addMouseWheelListener(layer, mouseWheelListener)
            scene.addDragListener(layer, dragListener)
        }

        override fun onRemove() {
            scene.removeMouseWheelListener(mouseWheelListener)
            scene.removeDragListener(dragListener)
        }
    }

    inner class MyViewpointModifier(private val originX: Int, private val originY: Int) : DragEventHandler {
        override fun dragTo(x: Int, y: Int) {
            val viewRect = viewport.viewRect
            viewRect.x += originX - x
            viewRect.y += originY - y
            val o: Any = viewport.view
            (o as JComponent).scrollRectToVisible(viewRect)
        }
    }

    init {
        scene.addInitializer(MyInitializer())
    }
}
